funnel/views/project.py

```diff
@@ -129,23 +131,77 @@ def csv(self):
+
+            if not self.obj.timezone:
+                form.timezone.data = current_app.config.get('TIMEZONE')
+            return render_form(form=form, title=_("Edit project"), submit=_("Save changes"), autosave=True)
+        elif request.method == 'POST':
+            if 'autosave' in request.form and request.form['autosave'] == 'true':
+                if 'revision' not in request.form:
+                    return {'error': _("Form must contain a valid revision ID.")}, 400
+
+                # ensure that the revision ID is valid
+                try:
+                    client_revision = UUID(request.form['revision']) if request.form['revision'] else None
+                except Exception as e:
+                    return {'error': _("Invalid UUID: {0!r}".format(e))}, 400
+
+                # find the last draft
+                draft = Draft.query.filter_by(table=Project.__tablename__, table_row_id=self.obj.uuid).order_by(Draft.updated_at.desc()).first()
+
+                if draft is not None:
+                    if client_revision is None or (client_revision is not None and draft.revision != client_revision):
+                        # draft exists, but the form did not send a revision ID,
+                        # OR revision ID sent by client does not match the last revision ID
+                        return {'error': _("There has been changes to this draft since you last edited it. Please reload.")}, 400
+                    elif client_revision is not None and draft.revision == client_revision:
+                        # revision ID sent my client matches, save updated draft data and update revision ID
+                        draft.body = {'form': request.form}
+                        draft.revision = uuid4()
+                else:
+                    # no draft exists, create one
+                    draft = Draft(table=Project.__tablename__, table_row_id=self.obj.uuid, body={'form': request.form}, revision=uuid4())
+                db.session.add(draft)
+                db.session.commit()
+                return {'draft': draft.body['form'], 'revision': draft.revision}
+            else:
+                if self.obj.parent_project:
+                    form = SubprojectForm(obj=self.obj, model=Project)
+                else:
+                    form = ProjectForm(obj=self.obj, parent=self.obj.profile, model=Project)
+                form.parent_project.query = Project.query.filter(Project.profile == self.obj.profile, Project.id != self.obj.id, Project.parent_project == None)  # NOQA
+                if form.validate_on_submit():
+                    form.populate_obj(self.obj)
+                    db.session.commit()
+                    flash(_("Your changes have been saved"), 'info')
+                    tag_locations.queue(self.obj.id)
+
+                    # find and delete drafts
+                    Draft.query.filter_by(table=Project.__tablename__, table_row_id=self.obj.uuid).delete()
```

Comment:
Once the columns are part of a primary key, this can change to `Draft.query.get((Project.__tablename__, self.obj.uuid)).delete()`.